### PR TITLE
feat(analytics): add finance chat evaluations

### PR DESCRIPTION
**Pull request by hiroppy**
# Summary

- Add a promptfoo evaluation harness for the production finance-chat tools with five deterministic demo-data scenarios.
- Validate answer grounding, cards, routes, tool ordering, dates, amounts, proportions, transaction attributes, and unsafe links with fail-closed assertions.
- Harden streaming/final link sanitization and deterministic demo-fixture verification based on review findings from the closed prior attempt.

## Linear Issue

- [HIR-110](https://linear.app/hiroppy/issue/HIR-110/llm%E3%81%AE%E8%A9%95%E4%BE%A1%E5%9F%BA%E7%9B%A4%E3%82%92%E4%BD%9C%E3%81%A3%E3%81%A6%E6%9C%9F%E5%BE%85%E5%80%A4%E9%80%9A%E3%82%8A%E3%81%AE%E8%BF%94%E7%AD%94%E3%81%8B%E3%83%86%E3%82%B9%E3%83%88%E3%81%99%E3%82%8B)

### Selected Quality Characteristics

| Quality characteristic | Why it is relevant | Acceptance criteria |
| ---------------------- | ------------------ | ------------------- |
| Functional suitability | Evaluation verdicts must match the finance-chat contract | Five scenarios validate expected text, cards, routes, and ordered tool evidence |
| Security | Model output must not expose unapproved links or unsupported claims | Unsafe/ambiguous links and malformed evidence fail closed before streaming |
| Reliability | Results must be repeatable and independent of personal data | Canonical demo fixture and no-cache CLI produce deterministic inputs |
| Maintainability | Review-driven edge cases must remain understandable and testable | Shared parsers/helpers have focused regression coverage and pass static checks |

### Test Techniques

| Test technique | Selection rationale | Expected coverage |
| -------------- | ------------------- | ----------------- |
| Equivalence partitioning | Separate valid/invalid evidence, links, claims, and card shapes | Accepted production responses and rejected unsupported variants |
| Boundary value analysis | Dates, month/year boundaries, zero/non-zero, ratios, and stream chunks are risk boundaries | Off-by-one periods, sign/denominator errors, and incomplete syntax |
| Decision table testing | Verdicts depend on claim, evidence order, scope, and output shape combinations | Fail-closed assertion branches and allowed equivalent card sets |
| State transition testing | Streaming sanitizer behavior changes as syntax arrives across chunks | No unsafe partial output before a link/anchor/code construct completes |
| Error guessing | LLM prose and Markdown have adversarial combinations not captured by simple schemas | Negation, paraphrases, nested syntax, IDN URLs, and malformed provenance |

### Primary Test Conditions

- Run all five promptfoo cases against the deterministic demo fixture.
- Accept only visible claims grounded by already-returned tool evidence with matching period and semantic label.
- Reject unauthorized URL forms across Markdown, HTML, autolinks, references, and streamed fragments.
- Reject malformed provider traces, fixture substitution, and ambiguous or contradictory financial claims.
- Preserve production finance-chat APIs and progressive streaming for safe text.

### Out-of-Scope Risks

- Live-model quality is not measured in this environment because AI provider credentials are unavailable; deterministic provider/assertion coverage and the CLI credential error path are verified.

## Verification

- [x] Unit tests
- [x] Type checking
- [x] Lint
- [x] Storybook tests
- [ ] E2E tests
- [x] Manual verification

### Commands Executed

```text
pnpm test — passed (analytics 725, web 860, db 320, all workspace tests)
pnpm turbo typecheck — passed (8 packages)
pnpm lint — passed
pnpm knip — passed
pnpm format:check — passed
pnpm --filter @mf-dashboard/analytics eval:chat — expected exit 100; cache disabled and all five missing-provider errors reported
```

### Checks Not Run

- E2E tests — not selected because the change is covered at the provider/assertion, API transform, and Storybook levels without changing an end-user interaction path.
- Live model evaluation — AI_PROVIDER, AI_MODEL, and AI_API_KEY are unavailable in this environment.

### CODE REVIEW

File: apps/web/src/app/api/chat/link-sanitizer.ts [hunk 1]
```diff
@@ -1,64 +1,98 @@
...
+      }
+      index += runLength - 1;
+      continue;
+    }
+
+    if (
+      !escaped &&
+      codeDelimiterLength === 0 &&
+      !htmlAnchorOpen &&
+      /^<a\b/iu.test(text.slice(index)) &&
```
Comment (hiroppy) [inline apps/web/src/app/api/chat/link-sanitizer.ts:111]: `splitCompleteFinanceChatText()` only treats partial `<a` input as pending when the current buffer already has a complete opening tag, an attribute followed by `=`, or exactly `<a`. A chunk ending at `<a title\n` hits the newline boundary and is emitted unsanitized; the following chunk ` href="javascript:alert(1)">click</a>` is then emitted as ` href="javascript:alert(1)">click`, so the client-visible stream reconstructs `<a title\n href="javascript:alert(1)">click`. Please keep any `<a` opening tag pending until `>` (or a safe self-close) is seen, regardless of partially typed attributes, and add adversarial chunk-boundary tests.

Comment (hiroppy) [reply apps/web/src/app/api/chat/link-sanitizer.ts:111]: ご指摘ありがとうございます。未完のraw anchor opener判定を、`<a`直後にASCII属性名が始まった時点で有効にし、属性の`=`有無に依存せず`>`またはself-closeまでbufferするよう修正しました。既存の`<a\\n`に加え、指摘ケースの`<a title\\n`でも最初のchunkが一切送信されず、後続の危険hrefを含むanchor全体がsanitizeされて`click。`だけになる回帰を追加しました。web unit 510件・full 861件を含む全テスト、typecheck、lint、knip、format checkが成功しています。

File: apps/web/src/app/api/chat/link-sanitizer.ts [hunk 2]
```diff
@@ -102,13 +149,25 @@ export function createFinanceChatLinkSanitizer<TOOLS extends ToolSet>(
...
+    const referenceDefinitionsById = new Map<string, Map<string, string>>();
 
     const enqueueSanitizedText = (
       controller: TransformStreamDefaultController,
       id: string,
       text: string,
     ) => {
-      const sanitizedText = sanitizeFinanceChatLinks(text, allowedHrefs);
+      const definitions = referenceDefinitionsById.get(id) ?? new Map<string, string>();
+      for (const match of text.matchAll(
```
Comment (hiroppy) [inline apps/web/src/app/api/chat/link-sanitizer.ts]: ここで `text.matchAll(...)` をそのまま走らせているため、Markdown の fenced/inline code 内にある `[route]: /group-a/cf/2026-07` も実際の reference definition として保存されます。その後の通常文 `[詳細][route]` が、コードブロック内で例示しただけの定義を使って `[詳細](/group-a/cf/2026-07)` に変換されてしまいます。手元でも ` ```\n[route]: /group-a/cf/2026-07\n```\n` の後に `[詳細][route]\n` を流すとリンク化されることを確認しました。reference definition の検出・保持はコード範囲を除外するか、既存の `sanitizeFinanceChatLinks` 側と同じ code masking を適用してください。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  for (let cursor = index - 1; cursor >= 0 && text[cursor] === "\\"; cursor -= 1) {
+    backslashCount += 1;
+  }
+  return backslashCount % 2 === 1;
+}
+
+function findMarkdownCodeRanges(text: string): MarkdownCodeRange[] {
+  const ranges: MarkdownCodeRange[] = [];
+  for (let index = 0; index < text.length; index += 1) {
+    if (text[index] !== "`" || isEscaped(text, index)) continue;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:144]: ご指摘ありがとうございます。CommonMarkのtilde fenceをopening 3文字以上・closingはopening以上の長さ・最大3空白indent・closing後は空白のみの規則でコード範囲としてmaskし、URL／email／raw anchor例を改変・収集しないよう修正しました。stream splitterにもtilde fence状態を追加し、chunkを跨ぐ`~~~`/`~~~~`で原文保持する回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+  const withoutHtmlLinks = sanitizeRawHtmlAnchors(masked, allowedHrefs);
+  const referenceDefinitions = new Map(
+    Array.from(
+      withoutHtmlLinks.matchAll(/^[ \t]*\[([^\]]+)\]\s*:\s*([^\s]+)(?:[ \t]+[^\r\n]*)?$/gimu),
+      ([, id, destination]) => [id.toLowerCase(), destination] as const,
+    ),
+  );
+  const withoutInvalidMarkdownLinks = sanitizeMarkdownInlineLinks(withoutHtmlLinks, allowedHrefs);
+  const withoutReferenceLinks = withoutInvalidMarkdownLinks.replace(
+    /(!?)\[([^\]]+)\]\[([^\]]+)\]/gu,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。reference parserをfull・collapsed・shortcutの3形式へ拡張し、collapsed/shortcutではlabelを正規化IDとしてallowlisted定義をinline linkへ解決するよう修正しました。stream splitterも両形式を定義到着までbufferします。sanitize 2形式・stream 2形式の回帰を追加し、analytics 728件・web full 864件を含む全ゲート成功を確認済みです。

File: apps/web/src/app/api/chat/link-sanitizer.ts (hunk 2 above)
Comment (hiroppy) [reply apps/web/src/app/api/chat/link-sanitizer.ts]: ご指摘ありがとうございます。reference definition抽出をanalyticsの共有helperへ集約し、backtick/tildeのinline・fenced codeをmaskしてから定義行だけを返すよう修正しました。stream側はこのhelperの結果だけを保持するため、コードブロック内の`[route]: ...`後に通常文`[詳細][route]`が来てもリンク化せず`詳細`として出力します。指摘どおりのtransform回帰を追加し、analytics 728件・web full 865件を含む全テスト、typecheck、lint、knip、format checkの成功を確認済みです。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    let matched = false;
+    for (
+      let closingStart = index + delimiterLength;
+      closingStart < text.length;
+      closingStart += 1
+    ) {
+      if (text[closingStart] !== "`" || isEscaped(text, closingStart)) continue;
+      let closingLength = 1;
+      while (text[closingStart + closingLength] === "`") closingLength += 1;
+      if (closingLength === delimiterLength) {
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:156]: ご指摘ありがとうございます。backtick fenceをinline code spanより先にline-orientedで検出し、3文字以上のopeningに対して同数以上のclosing runを受理するよう修正しました。stream splitterもbacktick fence状態を独立管理し、3→4 backtickのchunk回帰でURL／anchor例の原文保持を確認しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+export function findFinanceChatReferenceDefinitions(text: string): Array<{
+  destination: string;
+  id: string;
+  source: string;
+}> {
+  const { masked } = maskMarkdownCode(text);
+  return Array.from(
+    masked.matchAll(/^[ \t]*\[([^\]]+)\]\s*:\s*([^\s]+)(?:[ \t]+[^\r\n]*)?$/gimu),
+    ([source, id, destination]) => ({ destination, id: id.toLowerCase(), source }),
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。reference label正規化を共有helperへ集約し、trim・内部whitespace collapse・lowercaseをdefinition、explicit ID、collapsed/shortcut label、stream known IDすべてへ適用しました。`route label`と`route   label`を同一視するsanitize／stream回帰を追加し、analytics 730件・web full 867件を含む全ゲート成功を確認済みです。

File: packages/analytics/evals/assertions.ts [hunk 7]
```diff
@@ -0,0 +1,3249 @@
...
+      availableRows.filter(
+        (transaction) =>
+          typeof transaction === "object" &&
+          transaction !== null &&
+          "description" in transaction &&
+          normalize(String(transaction.description)) === normalize(description),
+      );
+    const mismatches: string[] = [];
+    for (const match of text.matchAll(
+      /(?:^|[。！？\n])\s*([^。！？\n]{1,80}?)のカテゴリ(?:は|が)\s*([^。！？\n]+?)(?:です|でした|である|だ)(?=[。！？\n]|$)/gu,
```
Comment (hiroppy) [inline packages/analytics/evals/assertions.ts:2901]: `requireTransactionToolGrounding` still false-passes transaction attribute claims that refer to rows by position/coreference instead of repeating the literal description. I reproduced an output with the correct retrieved rows/cards but text `前者は食費、後者は水道・光熱費です。` (swapping the two categories); `assertFinanceResponse()` returned `pass: true`. This lets the eval harness accept a response that misstates which transaction has which category. Consider adding adversarial tests for positional/pronominal references or rejecting unsupported transaction-attribute prose that cannot be bound to a retrieved row.

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+  let opening: RegExpExecArray | null;
+  while ((opening = openingPattern.exec(text)) !== null) {
+    if (marker === "`" && opening[2].includes("`")) continue;
+    const closingPattern = new RegExp(
+      `^ {0,3}${marker}{${opening[1].length},}[ \\t]*(?:\\r?$)`,
+      "gmu",
+    );
+    closingPattern.lastIndex = openingPattern.lastIndex;
+    const closing = closingPattern.exec(text);
+    if (closing === null) continue;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。閉じfenceが見つからない場合は有効なopeningからEOFまでをcode範囲としてmaskするよう修正しました。backtick/tilde両方の未終端fenceについて、URL・emailを原文保持し未承認リンクとして収集しない回帰を追加しています。analytics 736件を含む全テストと静的検査が成功しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+    ranges.push({ start: opening.index, end });
+    openingPattern.lastIndex = end;
+  }
+  return ranges;
+}
+
+function findMarkdownCodeRanges(text: string): MarkdownCodeRange[] {
+  const fenceRanges = [...findFencedCodeRanges(text, "`"), ...findFencedCodeRanges(text, "~")].sort(
+    (left, right) => left.start - right.start,
+  );
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。backtick/tildeの検出結果を開始位置順に統合し、既に選択した外側fenceと重なる内側rangeを除外してからmaskするよう修正しました。外側backtick内のtilde風行を含む指摘ケースで原文が一度だけ復元される回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+  label: string;
+  start: number;
+}
+
+function findMarkdownInlineLinks(text: string): MarkdownInlineLink[] {
+  const links: MarkdownInlineLink[] = [];
+  for (let index = 0; index < text.length; index += 1) {
+    const image = text[index] === "!" && text[index + 1] === "[";
+    const labelStart = image ? index + 1 : index;
+    if (text[labelStart] !== "[" || isEscaped(text, labelStart)) continue;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。image marker判定にも共有isEscapedのバックスラッシュ偶奇判定を適用しました。`\\詳細`をliteral `!`＋通常のallowlisted linkとして保持するsanitize/collect回帰を追加しています。

File: packages/analytics/evals/cases.yaml [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+    evaluationDate: "2026-07-31T03:00:00.000Z"
+  assert:
+    - type: is-json
+    - type: not-contains
+      value: データを取得できません
+    - type: javascript
+      value: file://./assertions.ts
+      config:
+        allowedVisibleAmounts: [5683100]
+        allowedVisibleDates: ["2026-07-31"]
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/evals/cases.yaml]: ご指摘ありがとうございます。総資産ケースの許可日をfixture唯一のsnapshot日`2026-07-24`へ修正し、取得証拠より新しい`2026-07-31`を受理しない契約に揃えました。全評価assertion回帰が成功しています。

File: packages/analytics/evals/provider.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+            : undefined,
+        );
+        if (route.success && (route.data === groupHref || route.data.startsWith(`${groupHref}/`))) {
+          allowedHrefs.add(route.data);
+        }
+      }
+      if (toolName === "presentFinanceCards") {
+        presentations.push(output);
+        allowedHrefsAtPresentation = allowedHrefsBeforeStep;
+        dataToolResultsAtPresentation = dataToolResultsBeforeStep;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/evals/provider.ts:168]: ご指摘ありがとうございます。`presentFinanceCards`の出力を再帰走査して全string fieldを`collectFinanceChatLinks`へ通し、allowlist外をpresentation時点で記録するよう修正しました。titleのwww、descriptionのbare email、hrefの外部URLを同時に検出する回帰を追加しています。

File: packages/analytics/evals/assertions.ts (hunk 7 above)
Comment (hiroppy) [reply packages/analytics/evals/assertions.ts:2901]: ご指摘ありがとうございます。literal descriptionへ安全に束縛できない前者・後者・前後の明細・1/2つ目等の位置照応による属性断定をfail-closedで拒否するよう修正しました。指摘どおり2行を取得して`前者は食費、後者は水道・光熱費です。`とする回帰で`未対応の明細照応`として失敗することを確認しています。

File: packages/analytics/evals/provider.ts [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+            ? output.href
+            : undefined,
+        );
+        if (route.success && (route.data === groupHref || route.data.startsWith(`${groupHref}/`))) {
+          allowedHrefs.add(route.data);
+        }
+      }
+      if (toolName === "presentFinanceCards") {
+        unauthorizedLinks.push(
+          ...collectVisibleStrings(output).flatMap((value) =>
```
Comment (hiroppy) [inline packages/analytics/evals/provider.ts:160]: `collectVisibleStrings(output)` feeds card output into `collectFinanceChatLinks()`, but that collector does not treat bare internal route strings like a card `href: "/other-group/cf/2026-07"` or `action.href` as links. As a result, `unauthorizedLinks` stays empty for fabricated/cross-group structured-card hrefs unless a specific case also has `expectedRoute`, leaving the eval harness able to false-pass card CTA provenance. Validate `href`/`action.href` fields directly against the allowed hrefs captured before `presentFinanceCards` instead of relying on Markdown/URL extraction.

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+    }
+    const end = closing.index + closing[0].length;
+    ranges.push({ start: opening.index, end });
+    openingPattern.lastIndex = end;
+  }
+  return ranges;
+}
+
+function findMarkdownCodeRanges(text: string): MarkdownCodeRange[] {
+  const fenceRanges = [...findFencedCodeRanges(text, "`"), ...findFencedCodeRanges(text, "~")]
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。4-spaceまたはtabで始まる各行をindented code rangeとしてfenced codeと同じ重複除去後にmaskするよう修正しました。URLとbare emailを含む両indent形式でsanitize/collectが原文保持する回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+      id,
+      destination,
+    ]),
+  );
+  const withoutInvalidMarkdownLinks = sanitizeMarkdownInlineLinks(withoutHtmlLinks, allowedHrefs);
+  const withoutReferenceLinks = withoutInvalidMarkdownLinks.replace(
+    /(!?)\[([^\]]+)\](?:\[([^\]]*)\])?(?!\s*[:(])/gu,
+    (match, imageMarker: string, label: string, explicitId: string | undefined) => {
+      if (imageMarker) return label;
+      const id = normalizeFinanceChatReferenceId(explicitId || label);
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:390]: ご指摘ありがとうございます。reference-style側でも`!`位置へ共有isEscapedを適用し、escaped markerはliteral `!`を残したまま通常のallowlisted reference linkへ解決するよう修正しました。指摘ケースの回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+}
+
+export function sanitizeFinanceChatLinks(text: string, allowedHrefs: Set<string>): string {
+  const { masked, restore } = maskMarkdownCode(text);
+  const withoutHtmlLinks = sanitizeRawHtmlAnchors(masked, allowedHrefs);
+  const referenceDefinitions = new Map(
+    findFinanceChatReferenceDefinitions(withoutHtmlLinks).map(({ id, destination }) => [
+      id,
+      destination,
+    ]),
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。static sanitizerとstream cacheの両方をfirst-write-winsへ変更しました。同一IDをcash-flow後にbalance-sheetで再定義しても最初のcash-flow routeへ解決するstatic/stream回帰で確認しています。

File: packages/analytics/evals/provider.ts (hunk 13 above)
Comment (hiroppy) [reply packages/analytics/evals/provider.ts:160]: ご指摘ありがとうございます。card output内の全`href` keyを再帰抽出し、presentation step開始前のallowlist snapshotへ直接完全一致検証するよう修正しました。cross-group card hrefと未取得action.hrefがunauthorizedLinksへ入る回帰を追加しています。analytics 741件、web full 867件を含む全ゲートが成功しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+function splitBareUrl(url: string) {
+  const match = /^(.*?)([.,!?;:。、，！？；：]+)$/u.exec(url);
+  let destination = match?.[1] ?? url;
+  let trailingText = match?.[2] ?? "";
+  const adjacentJapaneseTextAfterIdn = new RegExp(
+    `^((?:https?:\\/\\/|\\/\\/)(?:[^\\s./]+\\.)+[A-Za-z0-9-]+(?:[/?#][A-Za-z0-9\\-._~:/?#[\\]@!$&'*+,;=%]*)?)([\\p{Script=Hiragana}\\p{Script=Katakana}\\p{Script=Han}]+)$`,
+    "iu",
+  ).exec(destination);
+  const adjacentJapaneseText = new RegExp(
+    `^((?:https?:\\/\\/|\\/\\/)[A-Za-z0-9\\-._~:/?#[\\]@!$&'*+,;=%]+)([\\p{Script=Hiragana}\\p{Script=Katakana}\\p{Script=Han}]+)$`,
```
Comment (hiroppy) [inline packages/analytics/src/chat/link-sanitizer.ts:33]: `splitBareUrl()` trims punctuation before it peels off adjacent Japanese text, but this regexp can keep a trailing `.` inside the destination while moving only the Japanese suffix to `trailingText`. I reproduced `sanitizeFinanceChatLinks("https://example.com/0/cf/2026-07.を見て", new Set(["/0/cf/2026-07"]))` returning `"を見て"`; the otherwise allowlisted route is dropped because the resolver sees `/0/cf/2026-07.`. Re-run punctuation trimming after separating adjacent Japanese text, or make the adjacent-text match exclude sentence punctuation so valid route links followed by Japanese prose keep the CTA.

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 18]
```diff
@@ -0,0 +1,424 @@
...
+  let backslashCount = 0;
+  for (let cursor = index - 1; cursor >= 0 && text[cursor] === "\\"; cursor -= 1) {
+    backslashCount += 1;
+  }
+  return backslashCount % 2 === 1;
+}
+
+function findFencedCodeRanges(text: string, marker: "`" | "~"): MarkdownCodeRange[] {
+  const ranges: MarkdownCodeRange[] = [];
+  const openingPattern = new RegExp(`^ {0,3}(${marker}{3,})([^\\r\\n]*)(?:\\r?\\n|$)`, "gmu");
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。blockquote反復prefixとlist marker prefixをfence opening/closingの共通containerとして認識し、stream splitterにも同じ規則を適用しました。`> ~~~`内URLを原文保持するsanitize/collectとstream回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 19]
```diff
@@ -0,0 +1,424 @@
...
+    }
+    const end = closing.index + closing[0].length;
+    ranges.push({ start: opening.index, end });
+    openingPattern.lastIndex = end;
+  }
+  return ranges;
+}
+
+function findIndentedCodeRanges(text: string): MarkdownCodeRange[] {
+  return Array.from(text.matchAll(/^(?: {4}|\t)[^\r\n]*(?:\r?\n|$)/gmu), (match) => ({
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。indented code開始を文頭または空行後に限定し、通常paragraph直後の4-space行はlive URLとしてsanitize/collectするよう修正しました。streamでは末尾改行を次行context確定までbufferし、chunk境界でもparagraph継続を誤maskしない回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 20]
```diff
@@ -0,0 +1,424 @@
...
+export function sanitizeFinanceChatLinks(text: string, allowedHrefs: Set<string>): string {
+  const { masked, restore } = maskMarkdownCode(text);
+  const withoutHtmlLinks = sanitizeRawHtmlAnchors(masked, allowedHrefs);
+  const referenceDefinitions = new Map<string, string>();
+  for (const { id, destination } of findFinanceChatReferenceDefinitions(withoutHtmlLinks)) {
+    if (!referenceDefinitions.has(id)) referenceDefinitions.set(id, destination);
+  }
+  const withoutInvalidMarkdownLinks = sanitizeMarkdownInlineLinks(withoutHtmlLinks, allowedHrefs);
+  const withoutReferenceLinks = withoutInvalidMarkdownLinks.replace(
+    /(!?)\[([^\]]+)\](?:\[([^\]]*)\])?(?!\s*[:(])/gu,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:385]: ご指摘ありがとうございます。reference置換時に実際のopening bracket位置へ共有isEscapedを適用しました。`\\[詳細]`をliteralのまま保持し、定義だけを除去してdestination syntaxを付加しない回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 21]
```diff
@@ -0,0 +1,424 @@
...
+      if (content[offset] === "\\") offset += 1;
+      else if (content[offset] === "(") nestedDepth += 1;
+      else if (content[offset] === ")") nestedDepth -= 1;
+      else if (/\s/u.test(content[offset] ?? "") && nestedDepth === 0) {
+        splitIndex = offset;
+        break;
+      }
+    }
+    links.push({
+      destination: content.slice(0, splitIndex),
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。inline/referenceのdestination正規化helperを追加し、両端のangle bracketをallowlist照合とcollector返却前にunwrapするよう修正しました。指摘の`[詳細](</0/cf/2026-07>)`回帰でverified routeへ正規化されます。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 22]
```diff
@@ -0,0 +1,424 @@
...
+}
+
+export function findFinanceChatReferenceDefinitions(text: string): Array<{
+  destination: string;
+  id: string;
+  source: string;
+}> {
+  const { masked } = maskMarkdownCode(text);
+  return Array.from(
+    masked.matchAll(/^[ \t]*\[([^\]]+)\]\s*:\s*([^\s]+)(?:[ \t]+[^\r\n]*)?$/gimu),
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。reference definition parserを共有patternへ集約し、次行の1〜3 space付きquoted/parenthesized titleをdefinition sourceへ含めて一括除去するよう修正しました。continuation titleが可視文として残らない回帰を追加しています。analytics 746件・web full 869件を含む全ゲートが成功しています。

File: packages/analytics/src/chat/link-sanitizer.ts (hunk 17 above)
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:33]: ご指摘ありがとうございます。日本語suffix分離後のdestinationへ句読点分離を再適用し、句読点はsuffix側へ順序を保って戻すよう修正しました。指摘ケースで`/0/cf/2026-07.を見て`としてallowlisted CTAと本文を保持する回帰を追加し、analytics 747件・web full 869件を含む全ゲートが成功しています。

File: apps/web/src/app/api/chat/link-sanitizer.ts [hunk 23]
```diff
@@ -1,64 +1,149 @@
...
-    const escaped = index > 0 && text[index - 1] === "\\";
+    const escaped = isEscaped(text, index);
+    const lineStart = text.lastIndexOf("\n", index - 1) + 1;
+    const linePrefix = text.slice(lineStart, index);
+
+    if (
+      !escaped &&
+      character === "`" &&
+      codeDelimiterLength === 0 &&
+      fenceContainerPrefix.test(linePrefix)
```
Comment (hiroppy) [inline apps/web/src/app/api/chat/link-sanitizer.ts:58]: `splitCompleteFinanceChatText` starts tracking a backtick fence even while it is already inside a `~~~` fence, because this branch does not require `tildeFenceLength === 0`. If a tilde-fenced block contains an unmatched ``` literal, `backtickFenceLength` remains active after the `~~~` closes and subsequent prose is never emitted until `text-end`, defeating streaming. Add `tildeFenceLength === 0` to the backtick-fence branch and cover `~~~...``` literal...~~~\n次の文です。` as a regression.

Comment (hiroppy) [reply apps/web/src/app/api/chat/link-sanitizer.ts:58]: ご指摘ありがとうございます。backtick fence開始条件へ`tildeFenceLength === 0`を追加し、tilde fence内の```風literalが外側状態を汚染しないよう修正しました。外側close後の通常文まで即時completeになる回帰を追加しています。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 24]
```diff
@@ -0,0 +1,441 @@
...
+    }
+    const end = closing.index + closing[0].length;
+    ranges.push({ start: opening.index, end });
+    openingPattern.lastIndex = end;
+  }
+  return ranges;
+}
+
+function findIndentedCodeRanges(text: string): MarkdownCodeRange[] {
+  return Array.from(text.matchAll(/^(?:(?: {4}|\t)[^\r\n]*(?:\r?\n|$))+/gmu), (match) => ({
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。blockquote prefix後の4-space/tab indented blockをcode rangeへ追加し、URLをsanitize/collect対象にしない回帰を追加しました。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 25]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+function collectMarkdownInlineLinkDestinations(text: string): string[] {
+  return findMarkdownInlineLinks(text).flatMap(({ destination, label }) => [
+    destination,
+    ...collectMarkdownInlineLinkDestinations(label),
+  ]);
+}
+
+const referenceDefinitionPattern = String.raw`^[ \t]*\[([^\]]+)\]\s*:\s*([^\s]+)(?:[ \t]+[^\r\n]*)?(?:\r?\n[ \t]{1,3}(?:"[^"\r\n]*"|'[^'\r\n]*'|\([^\r\n)]*\))[ \t]*(?=\r?$))?`;
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts]: ご指摘ありがとうございます。reference definition共通patternへblockquote/list container prefixを追加しました。blockquote内のuse/definitionがallowlisted linkへ解決される回帰を追加しています。

File: apps/web/src/app/api/chat/link-sanitizer.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/web/src/app/api/chat/link-sanitizer.ts:58]: 同一原因のご指摘ありがとうございます。backtick fence branchをtilde fence外に限定し、mixed-fence回帰で外側close後のstream再開を確認しました。

File: packages/analytics/src/chat/link-sanitizer.ts [hunk 26]
```diff
@@ -0,0 +1,441 @@
...
+      return restored;
+    },
+  };
+}
+
+function findRawHtmlAnchors(text: string): RawHtmlAnchor[] {
+  const anchors: RawHtmlAnchor[] = [];
+  const openingPattern = /<a\b/giu;
+  let openingMatch: RegExpExecArray | null;
+  while ((openingMatch = openingPattern.exec(text)) !== null) {
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/analytics/src/chat/link-sanitizer.ts:198]: ご指摘ありがとうございます。escaped raw anchor全体を保護rangeとしてmaskし、raw anchor・bare URL両scannerから除外するよう修正しました。指摘例を原文保持しunauthorizedLinksへ収集しない回帰を追加しています。

File: apps/web/src/app/api/chat/link-sanitizer.ts [hunk 27]
```diff
@@ -82,13 +167,46 @@ export function splitCompleteFinanceChatText(text: string): {
...
+        continue;
+      lastBoundary = index;
+    }
   }
 
   if (lastBoundary < 0) return { complete: "", pending: text };
 
+  const complete = text.slice(0, lastBoundary + 1);
+  const referenceIds = Array.from(
+    complete.matchAll(/(?<!!)\[([^\]]+)\](?:\[([^\]]*)\])?(?!\s*[:(])/gu),
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/web/src/app/api/chat/link-sanitizer.ts]: ご指摘ありがとうございます。stream reference抽出をmarker＋bracketのescape parity判定へ揃え、escaped `!`後のcollapsed/shortcut referenceはdefinition到着までbufferする回帰を追加しました。analytics 750件・web full 871件を含む全ゲートが成功しています。

**Comment by hiroppy** [comment]
Closing this attempt to reset the evaluation design around final-response quality, per the HIR-110 rework direction.
